### PR TITLE
Refactor schema gen to generically create discriminated subschemas

### PR DESCRIPTION
**Pull request by jander-msft**
This change allows adding more discriminated subschemas in the future without reimplementing the concept for each set of options. Also, generate schemas from type information and cache it in a resolver so that type schemas can be reused across multiple type generations.

### CODE REVIEW

File: src/Tests/Microsoft.Diagnostics.Monitoring.ConfigurationSchema/SchemaGenerator.cs
Comment (kkeirstead) [inline src/Tests/Microsoft.Diagnostics.Monitoring.ConfigurationSchema/SchemaGenerator.cs:140]: Supposedly JsonSchemaResolver is now considered obsolete; does this change anything?

File: src/Tests/Microsoft.Diagnostics.Monitoring.ConfigurationSchema/SchemaGenerator.cs
Comment (kkeirstead) [inline src/Tests/Microsoft.Diagnostics.Monitoring.ConfigurationSchema/SchemaGenerator.cs:101]: I'm assuming these names should be `discriminating` instead of `descriminating`

File: src/Tests/Microsoft.Diagnostics.Monitoring.ConfigurationSchema/SchemaGenerator.cs
Comment (jander-msft) [reply src/Tests/Microsoft.Diagnostics.Monitoring.ConfigurationSchema/SchemaGenerator.cs:140]: Where do you see that it is obsolete? There's no Obsolete attribute on it, no compiler warnings/errors, and I haven't seen that in documentation.

Comment (kkeirstead) [reply src/Tests/Microsoft.Diagnostics.Monitoring.ConfigurationSchema/SchemaGenerator.cs:140]: Maybe I'm looking at the wrong thing? This is what I found when I searched: https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_Schema_JsonSchemaResolver.htm (it says at the top `Note: This API is now obsolete`).

Comment (jander-msft) [reply src/Tests/Microsoft.Diagnostics.Monitoring.ConfigurationSchema/SchemaGenerator.cs:140]: > Maybe I'm looking at the wrong thing? This is what I found when I searched: https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_Schema_JsonSchemaResolver.htm (it says at the top `Note: This API is now obsolete`).

The schema generation uses [NJsonSchema](https://github.com/RicoSuter/NJsonSchema), not Newtonsoft.Json. While NJsonSchema uses NewtonsoftJson as a dependency, this schema generation uses the NJsonSchema API.

File: src/Tests/Microsoft.Diagnostics.Monitoring.ConfigurationSchema/SchemaGenerator.cs
Comment (kkeirstead) [inline src/Tests/Microsoft.Diagnostics.Monitoring.ConfigurationSchema/SchemaGenerator.cs:108]: Not sure if this is actually a problem, but I noticed that you separated `FormatterName` as a RequiredProperty (line 69) from where you add `discriminatedPropertyName` as a property down below. In theory, if `subSchema` is null, then when we initialize `subSchema` we won't have `FormatterName` as a required property. Do you think this is a problem? Should we add `subSchema.RequiredProperties.Add(nameof(ConsoleLoggerOptions.FormatterName))` on line 120?

File: src/Tests/Microsoft.Diagnostics.Monitoring.ConfigurationSchema/SchemaGenerator.cs
Comment (kkeirstead) [inline src/Tests/Microsoft.Diagnostics.Monitoring.ConfigurationSchema/SchemaGenerator.cs:111]: nit: Looks like you missed this `descriminating` (and the one on line 117)

File: src/Tests/Microsoft.Diagnostics.Monitoring.ConfigurationSchema/SchemaGenerator.cs
Comment (jander-msft) [reply src/Tests/Microsoft.Diagnostics.Monitoring.ConfigurationSchema/SchemaGenerator.cs:108]: I don't think that's going to be an issue, in fact, the "required" array on the sub schemas in this case might not actually be needed at all (for any of the given subschemas, if FormatterName is "const": "<somevalue>", then by virtue of the fact that it has a value then the requiredness of the property is fulfilled implicitly).

There will be other schemas where the properties in the parent schema are already defined to be required. It _might_ not be necessary to redeclare that those properties in the subschema are again required. My empirical testing has not shown that this is necessary.

**Review by kkeirstead — APPROVED** [review]
I think this is everything I noticed when going over the PR, so I can Approve this, but if you want to get another set of eyes on it before merging maybe @kelltrick can look over it as well.

**Pull request merged**
